department-of-veterans-affairs/caseflow
# Decision Review Queue In-Progress Tasks Query Optimization

Provides most of the functionality of [APPEALS-12465](https://vajira.max.gov/browse/APPEALS-12468) while being decoupled from the pagination effort. This is an "immediate relief package" that we plan on delivering to stakeholders while we wrap up our pagination implementation.

### Description
This story refactors the `in_progress_tasks` method used to gather open business line tasks for decision review queues to utilize the hybrid ActiveRecord-Arel query that was detailed in @TylerBroyles's [research](https://hackmd.io/5rfIor7bTtqLOeU6gFGmpA#Implementing-the-query-through-a-hybrid-approach-of-Rails-and-Arel).

By converting the query method to one that offloads all task filtering to the database server (instead of having Ruby do the heavy-lifting) we have observed **tremendous** improvements in the time it takes to acquire open decision review tasks for a particular business line.

### Acceptance Criteria
- [ ] The query that populates our decision reviews within the queue is optimized and there is a tangible (benchmarked) improvement upon the both the performance of the query and the loading time of the queue
- [ ] Query is extensible so that [APPEALS-12467](https://urldefense.com/v3/__https:/vajira.max.gov/browse/APPEALS-12467__;!!May37g!KjK5qZzC2SGINPnPmJGWNOekH3TrNLAdULdvZIOjv92aChxLaK0JwNj9az7xrIPmKXEPEpADm196tz-nOlGAxSDh0gvtIFKk$) (searching), [APPEALS-12468](https://urldefense.com/v3/__https:/vajira.max.gov/browse/APPEALS-12468__;!!May37g!KjK5qZzC2SGINPnPmJGWNOekH3TrNLAdULdvZIOjv92aChxLaK0JwNj9az7xrIPmKXEPEpADm196tz-nOlGAxSDh0ul4lNWL$) (filtering), & [APPEALS-12530](https://urldefense.com/v3/__https:/vajira.max.gov/browse/APPEALS-12530__;!!May37g!KjK5qZzC2SGINPnPmJGWNOekH3TrNLAdULdvZIOjv92aChxLaK0JwNj9az7xrIPmKXEPEpADm196tz-nOlGAxSDh0ldYf95I$) (ordering) can be successfully implemented

### Testing Plan
1. Open the Rails console and disable the `:board_grant_effectuation_task` FeatureToggle:
```ruby
FeatureToggle.disable!(:board_grant_effectuation_task)
```

2. Create a myriad of tasks for the later testing steps. Feel free to adjust the number of tasks you will create, but note that later testing steps will reference the quantities created below:

```ruby
# Save the VHA business line to a variable for later use
vha = BusinessLine.find_by(url: "vha")

# Check count of open VHA business line tasks before proceeding:
vha.tasks.open.size

#
# Section 1
# Everything in the below section should always appear in the in_progress_tasks result set.
#
active_hlr_tasks = 100.times.map { FactoryBot.create(:higher_level_review_vha_task) }
active_sc_tasks = 100.times.map { FactoryBot.create(:supplemental_claim_vha_task) }
active_vrr_tasks = 100.times.map { FactoryBot.create(:veteran_record_request_task, assigned_to: vha) }
active_vrr_tasks.each { |t| FactoryBot.create(:request_issue, decision_review: t.appeal) }

#
# Section 2
# Everything in the below section should never appear in the in_progress_Tasks result set.
#
completed_hlr_tasks = 100.times.map { FactoryBot.create(:higher_level_review_vha_task) }.each(&:completed!)
completed_sc_tasks = 100.times.map { FactoryBot.create(:supplemental_claim_vha_task) }.each(&:completed!)
completed_vrr_tasks = 100.times.map { FactoryBot.create(:veteran_record_request_task, assigned_to: vha) }.each(&:completed!)
open_vrr_tasks_without_ris = 100.times.map { FactoryBot.create(:veteran_record_request_task, assigned_to: vha) }

#
# Section 3
# These BoardGrantEffectuationTasks (BGTs) will appear if the board_grant_effectuation_task
# FeatureToggle is enabled
#
bgt_tasks = 100.times.map { FactoryBot.create(:board_grant_effectuation_task, assigned_to: vha) }
bgt_tasks.each do |t|
  issue = FactoryBot.create(:request_issue, decision_review: t.appeal)
  issue.update!(closed_at: Time.zone.now, closed_status: "decided")
end
```

3. Check the number of open VHA tasks and ensure that the number has increased by 500:
`vha.tasks.open.size`

4. Now it's time to check on the actual query! Out of the 500 open VHA tasks, only 300 of these should appear in the result set of `.in_progress_tasks` at this time. Tasks from `open_vrr_tasks_without_ris` will not appear due to their lack of request issues on their decision reviews, and `bgt_tasks` will be omitted due to the `:board_grant_effectuation_task` FeatureToggle being disabled.

```ruby
vha.in_progress_tasks.group_by(&:type).map { |k,v| { type: k, num: v.size } }

=>
# Your totals may differ if you already had some tasks in your local DB
[{:type=>"VeteranRecordRequest", :num=>100}, {:type=>"DecisionReviewTask", :num=>200}]
```

5. Mark some HLR request issues as ineligible:
```ruby
active_hlr_tasks[0...50].each { |t| t.appeal.request_issues.each { |ri| ri.update!(ineligible_reason: "untimely") } }
```

6. Rerun the command from step 4. 50 fewer `DecisionReviewTasks` should appear:
```ruby
vha.in_progress_tasks.group_by(&:type).map { |k,v| { type: k, num: v.size } }

=>
[{:type=>"VeteranRecordRequest", :num=>100}, {:type=>"DecisionReviewTask", :num=>150}]
```

7. Mark some SC request issues as closed:
```ruby
active_sc_tasks[0...50].each do |t| 
  t.appeal.request_issues.each do |ri|
    ri.update!(closed_at: Time.zone.now, closed_status: "decided")
  end
end
```

8. Check the task type counts again, and look for another decrease of 50 for `DecisionReviewTasks`:
```ruby
vha.in_progress_tasks.group_by(&:type).map { |k,v| { type: k, num: v.size } }

=>
[{:type=>"VeteranRecordRequest", :num=>100}, {:type=>"DecisionReviewTask", :num=>100}]
```

9. Enable the `:board_grant_effectuation_task` FeatureToggle:
```ruby
FeatureToggle.enable!(:board_grant_effectuation_task)
```

10. Check the task type counts again. You should see 100 BoardGrantEffectuationTasks, and all other counts should remain the same:
```ruby
vha.in_progress_tasks.group_by(&:type).map { |k,v| { type: k, num: v.size } }

=>
[{:type=>"BoardGrantEffectuationTask", :num=>100},{:type=>"VeteranRecordRequest", :num=>100}, {:type=>"DecisionReviewTask", :num=>100}]
```

11. Complete some VeteranRecordRequest tasks:
```ruby
active_vrr_tasks[0...50].each(&:completed!)
```

12. Check the task type counts one last time. The `VeteranRecordRequest` quantity in the result set should be down by 50:
```ruby
vha.in_progress_tasks.group_by(&:type).map { |k,v| { type: k, num: v.size } }

=>
[{:type=>"VeteranRecordRequest", :num=>50}, {:type=>"DecisionReviewTask", :num=>100}, {:type=>"BoardGrantEffectuationTask", :num=>100}]
```

13. Something to note is the number of queries being performed each time `vha.in_progress_tasks` is invoked. In the console output above the command from step 12 you should see only **5 queries**. Prior to these changes, you would have seen hundreds.
```ruby
Task Load(X.Xms) # Primary query

Organization Load (X.Xms) # Eager-loaded association for Task's "assigned_to" attribute
Appeal Load (X.Xms) # Eager-loaded associated for Task's "appeal" attribute
SupplementalClaim Load (X.Xms) # Eager-loaded associated for Task's "appeal" attribute
HigherLevelReview Load (X.Xms) # Eager-loaded associated for Task's "appeal" attribute
```

Make sure that the number of milliseconds it takes to perform these queries is small (<30ms at least, but expect closer to 10ms whenever run locally). The same query prior to the refactoring that took place in this story would have taken several seconds to complete.
